src/rust/en.tsumino/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+		}
+	}
+	let total = json.get("pageCount").as_int().unwrap_or(0) as i32;
+
+	Ok(MangaPageResult {
+		manga: manga_arr,
+		has_more: page < total,
+	})
+}
+
+#[get_manga_listing]
+fn get_manga_listing(listing: Listing, page: i32) -> Result<MangaPageResult> {
+	let url = String::from("https://www.tsumino.com/search/operate/");
+	let mut parameters = String::new();
+	parameters.push_str("PageNumber=");
+	parameters.push_str(&encode_uri(page.to_string()));
+	parameters.push_str("&Sort=");
+	parameters.push_str(&encode_uri(listing.name));
+	let request = Request::new(&url, HttpMethod::Post)
+		.header("User-Agent", "Aidoku")
+		.body(parameters);
+	let json = request.json()?.as_object()?;
+	let data = json.get("data").as_array()?;
+
+	let mut manga_arr: Vec<Manga> = Vec::new();
+	for manga in data {
+		let obj = manga.as_object()?;
+		if let Ok(manga) = parser::parse_list(obj) {
+			manga_arr.push(manga);
+		}
+	}
+	let total = json.get("pageCount").as_int().unwrap_or(0) as i32;
+
+	Ok(MangaPageResult {
+		manga: manga_arr,
+		has_more: page < total,
+	})
+}
+
+#[get_manga_details]
+fn get_manga_details(id: String) -> Result<Manga> {
+	let url = format!("https://www.tsumino.com/entry/{}", id);
+	let request = Request::new(&url, HttpMethod::Get).header("User-Agent", "Aidoku");
+	let html = request.html()?;
+	let manga = parser::parse_manga(html)?;
+	Ok(Manga {
+		id,
+		title: manga.title,
+		cover: manga.cover,
+		author: manga.author,
+		description: manga.description,
+		url,
+		categories: manga.categories,
+		status: manga.status,
+		nsfw: manga.nsfw,
+		viewer: manga.viewer,
+		..Default::default()
+	})
+}
+
+#[get_chapter_list]
+fn get_chapter_list(id: String) -> Result<Vec<Chapter>> {
+	let url = format!("https://www.tsumino.com/entry/{}", id);
+	let request = Request::new(&url, HttpMethod::Get).header("User-Agent", "Aidoku");
+	let html = request.html()?;
+	let date_uploaded = html
+		.select("div.book-info-container #Uploaded")
+		.text()
+		.0
+		.as_date("yyyy MMMM d", Some("en_US"), None)
+		.unwrap_or(-1.0);
+	// skitty wants this to be put in
+	let scanlator = html
+		.select("div.book-info-container")
+		.select("#Uploader")
+		.select("a")
+		.array()
+		.get(0)
+		.as_node()
+		.expect("Failed to get uploader")
+		.text()
+		.read();
+
+	Ok(vec![Chapter {
+		id,
+		title: String::from("Chapter 1"),
+		volume: -1.0,
+		chapter: 1.0,
+		date_updated: date_uploaded,
+		url,
+		scanlator,
+		lang: String::from("en"),
+	}])
+}
+
+#[get_page_list]
+fn get_page_list(id: String, _: String) -> Result<Vec<Page>> {
+	let request = Request::new(
+		format!("https://www.tsumino.com/Read/Index/{}", id),
+		HttpMethod::Get,
+	)
+	.header("User-Agent", "Aidoku");
+	let mut pages: Vec<Page> = Vec::new();
+	let html = request.html()?;
+	let num_pages = html
+		.select("h1")
+		.text()
+		.read()
+		.split(' ')
+		.last()
+		.unwrap()
+		.parse::<i32>()
+		.ok();
+	match num_pages {
+		Some(num_pages) => {
+			for i in 1..(num_pages + 1) {
```

Comment:
could call this `index` instead of `i` so you can just do `index,` in the `Page` struct (minor complaint lol)